taichi/codegen/codegen_llvm_cuda.cpp

```diff
@@ -33,8 +29,10 @@ class CodeGenLLVMCUDA : public CodeGenLLVM {
 
   CodeGenLLVMCUDA(Kernel *kernel) : CodeGenLLVM(kernel) {
 #if defined(TI_WITH_CUDA)
-    cudaDeviceGetAttribute(&num_SMs, cudaDevAttrMultiProcessorCount, 0);
-    cudaDeviceGetAttribute(&max_block_dim, cudaDevAttrMaxBlockDimX, 0);
+    CUDADriver::get_instance().device_get_attribute(
```

Comment:
Sorry i should have pointed it out earlier. If you do not want to scatter `CUDADriver::get_instance()` all over the code base, one way to do it is to pass the pointer/reference of the global `CUDADriver` instance into the constructors of these classes (e.g. `CodeGenLLVMCUDA`, `Profiler`, etc).  This decouples the rest of the classes from the global instance. The benefits would be, say we have  a unit test and we only need to pass in a mock driver, you can easily substitute the real instance...

But I don't think we need to do this here...

Reply:
Good idea! I'll try to do this in a future PR.